File: 백준/Gold/25953. 템포럴 그래프/템포럴 그래프.py
# 템포럴 그래프
import sys
input = sys.stdin.readline

N, T, M = map(int,input().split())
S, E = map(int,input().split())
MAX = 10_000_001

# dp
dp = [[MAX] * N for _ in range(T+1)]
dp[0][S] = 0

for i in range(1, T+1):
    for j in range(N):
        dp[i][j] = dp[i-1][j]
        
    for j in range(M):
        n1, n2, w = map(int,input().split())
        dp[i][n1] = min(dp[i][n1], dp[i-1][n2] + w)
        dp[i][n2] = min(dp[i][n2], dp[i-1][n1] + w)
        
if dp[T][E] == MAX:
    print(-1)
else:
    print(dp[T][E])

File: 백준/Gold/25953. 템포럴 그래프/템포럴 그래프.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int T = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        int S = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());

        final int MAX = 10_000_001;

        // dp[n] -> N번째 지역까지 가기 위한 최소 비용
        int[] dp = new int[N];
        Arrays.fill(dp, MAX);
        dp[S] = 0; // 시작 지역의 비용은 0

        for (int i = 1; i < T+1; i++) {
            // 배열을 복사하여 이전 값을 사용
            int[] prev = Arrays.copyOf(dp, dp.length);

            for (int j = 0; j < M; j++) {
                st = new StringTokenizer(br.readLine());

                int n1 = Integer.parseInt(st.nextToken());
                int n2 = Integer.parseInt(st.nextToken());
                int w =  Integer.parseInt(st.nextToken());

                // 이전값과 비교하여 최소값으로 갱신
                dp[n1] = Math.min(dp[n1], prev[n2] + w);
                dp[n2] = Math.min(dp[n2], prev[n1] + w);
            }
        }

        if (dp[E] == MAX) {
            System.out.println(-1);
        }else {
            System.out.println(dp[E]);
        }
    }
}
